File: LeetCodeRecord/MethodFirst/56-Merge.java
package MethodFirst;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

class Merge {
    public int[][] merge(int[][] intervals) {
        List<int[]> merge = new ArrayList<int[]>();

        Arrays.sort(intervals, new Comparator<int[]>(){
            @Override
            public int compare(int[] interval1, int[] interval2) {
                return interval1[0] - interval2[0];
            }
        });

        for (int i = 0; i < intervals.length; i++) {
            int left = intervals[i][0];
            int right = intervals[i][1];

            if (i == 0 || merge.get(merge.size() - 1)[1] < left) {
                merge.add(intervals[i]);
            } else {
                merge.get(merge.size() - 1)[1] = Math.max(right, merge.get(merge.size() - 1)[1]);
            }
        }

        return merge.toArray(new int[merge.size()][]);
    }    
}

File: LeetCodeRecord/ACFirst-Golang/56-Merge.go
package main

import (
	"sort"
)

type NewInt [][]int

func (data NewInt) Len() int {
	return len(data)
}

func (data NewInt) Swap(i, j int) {
	data[i], data[j] = data[j], data[i]
}

func (data NewInt) Less(i, j int) bool {
	return data[i][0] < data[j][0]
}

func merge(intervals [][]int) [][]int {
	sort.Sort(NewInt(intervals))
	res := [][]int{}

	for i, interval := range intervals {
		left := interval[0]
		right := interval[1]

		if i == 0 || left > res[len(res)-1][1] {
			res = append(res, interval)
		} else if left <= res[len(res)-1][1] {
			res[len(res)-1][1] = max(right, res[len(res)-1][1])
		}
	}

	return res
}

func max(i, j int) int {
	if i > j {
		return i
	}

	return j
}
